cpp/src/arrow/util/async_generator_test.cc

```diff
@@ -67,14 +68,14 @@ class TrackingGenerator {
     return state_->source();
   }
 
-  int num_read() { return state_->num_read; }
+  int num_read() { return state_->num_read.load(); }
 
  private:
   struct State {
     explicit State(AsyncGenerator<T> source) : source(std::move(source)), num_read(0) {}
 
     AsyncGenerator<T> source;
-    int num_read;
+    std::atomic<int> num_read;
```

Comment:
@westonpace I had to change this as well, got a sporadic race condition in `MergedGeneratorTests/MergedGeneratorTestFixture.MergedLimitedSubscriptions/1`.